### PR TITLE
port: Better binary data support in Expression/LG functions

### LINKED ISSUE #3385: port: Better binary data support in Expression/LG functions (#5118)
The changes in [Better binary data support in Expression/LG functions (#5118)](https://github.com/microsoft/botbuilder-dotnet/pull/5118) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
fixes: #5117

### Support binary input for prebuilt function `string`
Now, `string` would support converting a binary object into a normal string with `UTF-8` decoding.
For example:
`string(binary('hello'))`  -> 'hello'

### Refine `fromFile` 
Originally, `fromFile` accepts only one parameter that refers to the file path. The step is as follows:
- Find file
- Read file with string stream and get the string result
- Evaluate the text

In this PR, we introduce the second parameter of `fromFile`. 
Its value could be: `evaluated (default)`, `raw` or `binary`

So the change of this function is:
`fromFile(path: string)` ->
`fromFile(path: string, format: 'evaluated'(default) | 'raw' | 'binary')`

#### `raw` format
In some situations, users do not want to evaluate the text, and just want to keep the original content text.

Example:
There is a file, xx.txt, content is:
```
hi ${name}
```

Original, user could use `fromFile('xx.txt')` to get `hi jack`, `hi Lucy`,etc. according to the memory/properties passed in. 
But there is no way to get the original text. 
Now, user could use `fromFile('xx.txt', 'raw')` to get `hi ${name}`

#### `binary` format
In another situation, the user wants to get the binary result of the file.

Example:
`string(fromFile('xx.txt', 'binary'))`  => 'hi ${name}'
If you want to evaluate it directly, just append `expandText` function:
`expandText(string(fromFileBinary('xx.txt')))`  => 'hi Jack', 'hi Lucy' ...

Another example for Bot builder:
For Bot response, if the user wants to send a local image to the client, here is the solution:
```
# attachmentTemplate
[Attachment
    name = myImageName
    contentType = image/png
    contentUrl = data:image/png;base64,${${base64(fromFile(xx/name.png, 'binary'))}}
]
```
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes #3385

### Support binary input for prebuilt function `string`
Now, `string` would support converting a binary object into a normal string with `UTF-8` decoding.
For example:
`string(binary('hello'))`  -> 'hello'

### Refine `fromFile` 
Originally, `fromFile` accepts only one parameter that refers to the file path. The step is as follows:
- Find file
- Read file with string stream and get the string result
- Evaluate the text

In this PR, we introduce the second parameter of `fromFile`. 
Its value could be: `evaluated (default)`, `raw` or `binary`

So the change of this function is:
`fromFile(path: string)` ->
`fromFile(path: string, format: 'evaluated'(default) | 'raw' | 'binary')`

#### `raw` format
In some situations, users do not want to evaluate the text, and just want to keep the original content text.

Example:
There is a file, xx.txt, content is:
```
hi ${name}
```

Original, user could use `fromFile('xx.txt')` to get `hi jack`, `hi Lucy`,etc. according to the memory/properties passed in. 
But there is no way to get the original text. 
Now, user could use `fromFile('xx.txt', 'raw')` to get `hi ${name}`

#### `binary` format
In another situation, the user wants to get the binary result of the file.

Example:
`string(fromFile('xx.txt', 'binary'))`  => 'hi ${name}'
If you want to evaluate it directly, just append `expandText` function:
`expandText(string(fromFileBinary('xx.txt')))`  => 'hi Jack', 'hi Lucy' ...

Another example for Bot builder:
For Bot response, if the user wants to send a local image to the client, here is the solution:
```
# attachmentTemplate
[Attachment
    name = myImageName
    contentType = image/png
    contentUrl = data:image/png;base64,${${base64(fromFile(xx/name.png, 'binary'))}}
]
```

### CODE REVIEW

**Comment by Danieladu** [comment]
Additional  change: Migrate all `myget` feed package to `yarnpkg` registry in `yarn.lock`

**Comment by Danieladu** [comment]
Another question: 
Here is the history: I want to drop the use of `Buffer`, instead with Third-party library "atob-lite" and "btoa-lite" and `TextDecoder/TextEncoder`.
Here are the concerns and thinking:
TextDecoder/TextEncoder is introduced into `node.js` during `v8.3`, but as a global variable in `v11.0`, so, in node 10.(Which we should support currently in CI pipeline.), May apply this code:
```
const util = require('util');
const encoder = new util.TextEncoder();
```
But in node >= 11 and browser(> IE 11), the below code is enough:
```
const encoder = new TextEncoder();
```
There are several options:
1. Dynamic check the environment and node version, to apply the implements for different platform and versions.
2. Upgrade our minimum supported node from 10 to 12， which is consistent with Composer.
3. Use some third-Party libraries, like [@sinonjs/text-encoding](https://www.npmjs.com/package/@sinonjs/text-encoding)
4. Keep Buffer operation.
@joshgummersall what do you think?

**Comment by joshgummersall** [comment]
It looks like `require('util').TextEncoder` still works in Node 12+

```
~ > node -v
v14.16.0
~ > node -e 'console.log(require("util").TextEncoder)'
[class TextEncoder]
```

I would just use that as it's backward-compatible.

File: libraries/botbuilder-lg/src/expander.ts [hunk 1]
```diff
@@ -657,19 +657,38 @@ export class Expander extends AbstractParseTreeVisitor<string[]> implements LGTe
     private readonly fromFile = (): any => (args: readonly any[]): any => {
         const filePath: string = TemplateExtensions.normalizePath(args[0].toString());
         const resourcePath: string = this.getResourcePath(filePath);
-        const stringContent = fs.readFileSync(resourcePath, 'utf-8');
+        let format = FileFormat.Evaluated;
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/src/expander.ts:660]:
```typescript
const expected = args[1].toLowerCase();
const format = Object.values(FileFormat).find(format => format.toLowerCase() === expected) ?? FileFormat.evaluated;
```

File: libraries/botbuilder-lg/src/evaluator.ts [hunk 2]
```diff
@@ -642,19 +664,38 @@ export class Evaluator extends AbstractParseTreeVisitor<any> implements LGTempla
     private readonly fromFile = (): any => (args: readonly any[]): any => {
         const filePath: string = TemplateExtensions.normalizePath(args[0].toString());
         const resourcePath: string = this.getResourcePath(filePath);
-        const stringContent = fs.readFileSync(resourcePath, 'utf-8');
+        let format = FileFormat.Evaluated;
+        if (args.length > 1) {
+            for (const buildinFormat of Object.values(FileFormat)) {
```
Comment (joshgummersall) [inline libraries/botbuilder-lg/src/evaluator.ts:669]: `.find` here too

**Review by joshgummersall — APPROVED** [review]

**Pull request merged**